File: webui/static/js/filters.js
window.FilterConfig = {
  media: [
    {
      key: "image",
      labelKey: "filters.media.image",
      patterns: [
        /^image_/i,
        /^icon_/i,
        /^spriteasset_/i,
        /^ui_/i,
        /^launcher_/i,
      ],
    },
    {
      key: "video",
      labelKey: "filters.media.video",
      patterns: [/\.usm$/i, /^music_lyric_video_/i, /^picture_/i],
    },
    {
      key: "audio",
      labelKey: "filters.media.audio",
      patterns: [/^bgm_/i, /^vo_/i, /^se_/i, /^music_/i, /\.acb$/i, /\.awb$/i],
    },
    {
      key: "model",
      labelKey: "filters.media.model",
      patterns: [/^3d_/i, /^ingame_/i, /\.playable\.assetbundle$/i],
    },
    {
      key: "motion",
      labelKey: "filters.media.motion",
      patterns: [/^mot_/i, /\.anim\.assetbundle$/i, /\.controller\.assetbundle$/i],
    },
    {
      key: "story",
      labelKey: "filters.media.story",
      patterns: [/^story_/i, /\.txt$/i, /^quest_/i, /^section_/i],
    },
    {
      key: "chart",
      labelKey: "filters.media.chart",
      patterns: [/^rhythmgame_/i, /^musicscore_/i, /\.bytes$/i, /\.csv$/i],
    },
  ],
  characters: [
    {
      key: "sachi",
      labelKey: "filters.character.sachi",
      patterns: [/(^|[_-])sachi([_.-]|$)/i],
    },
    {
      key: "kozue",
      labelKey: "filters.character.kozue",
      patterns: [/(^|[_-])kozue([_.-]|$)/i],
    },
    {
      key: "tsuzuri",
      labelKey: "filters.character.tsuzuri",
      patterns: [/(^|[_-])tsuzuri([_.-]|$)/i],
    },
    {
      key: "megumi",
      labelKey: "filters.character.megumi",
      patterns: [/(^|[_-])megumi([_.-]|$)/i],
    },
    {
      key: "kaho",
      labelKey: "filters.character.kaho",
      patterns: [/(^|[_-])kaho([_.-]|$)/i],
    },
    {
      key: "sayaka",
      labelKey: "filters.character.sayaka",
      patterns: [/(^|[_-])sayaka([_.-]|$)/i],
    },
    {
      key: "rurino",
      labelKey: "filters.character.rurino",
      patterns: [/(^|[_-])rurino([_.-]|$)/i],
    },
    {
      key: "ginko",
      labelKey: "filters.character.ginko",
      patterns: [/(^|[_-])ginko([_.-]|$)/i],
    },
    {
      key: "kosuzu",
      labelKey: "filters.character.kosuzu",
      patterns: [/(^|[_-])kosuzu([_.-]|$)/i],
    },
    {
      key: "hime",
      labelKey: "filters.character.hime",
      patterns: [/(^|[_-])hime([_.-]|$)/i],
    },
    {
      key: "ceras",
      labelKey: "filters.character.ceras",
      patterns: [/(^|[_-])ceras([_.-]|$)/i],
    },
    {
      key: "izumi",
      labelKey: "filters.character.izumi",
      patterns: [/(^|[_-])izumi([_.-]|$)/i],
    },
    {
      key: "tsukasa",
      labelKey: "filters.character.tsukasa",
      patterns: [/(^|[_-])tsukasa([_.-]|$)/i],
    },
    {
      key: "hiromi",
      labelKey: "filters.character.hiromi",
      patterns: [/(^|[_-])hiromi([_.-]|$)/i],
    },
  ],
  tags: [
    {
      key: "skill",
      labelKey: "filters.tag.skill",
      tokens: ["skill"],
      patterns: [/(^|[_-])skill([_.-]|$)/i],
    },
    {
      key: "middle",
      labelKey: "filters.tag.middle",
      tokens: ["middle"],
      patterns: [/(^|[_-])middle([_.-]|$)/i],
    },
    {
      key: "full",
      labelKey: "filters.tag.full",
      tokens: ["full"],
      patterns: [/(^|[_-])full([_.-]|$)/i],
    },
    {
      key: "half",
      labelKey: "filters.tag.half",
      tokens: ["half"],
      patterns: [/(^|[_-])half([_.-]|$)/i],
    },
    {
      key: "season",
      labelKey: "filters.tag.season",
      tokens: ["season"],
      patterns: [/(^|[_-])season([_.-]|$)/i],
    },
    {
      key: "adv",
      labelKey: "filters.tag.adv",
      tokens: ["adv"],
      patterns: [/(^|[_-])adv([_.-]|$)/i],
    },
    {
      key: "item",
      label: "item",
      tokens: ["item"],
      patterns: [/(^|[_-])item([_.-]|$)/i],
    },
    {
      key: "get",
      label: "get",
      tokens: ["get"],
      patterns: [/(^|[_-])get([_.-]|$)/i],
    },
    {
      key: "frame",
      label: "frame",
      tokens: ["frame"],
      patterns: [/(^|[_-])frame([_.-]|$)/i],
    },
    {
      key: "specialappeal",
      label: "specialappeal",
      tokens: ["specialappeal"],
      patterns: [/(^|[_-])specialappeal([_.-]|$)/i],
    },
    {
      key: "area",
      label: "area",
      tokens: ["area"],
      patterns: [/(^|[_-])area([_.-]|$)/i],
    },
    {
      key: "laugh",
      label: "laugh",
      tokens: ["laugh"],
      patterns: [/(^|[_-])laugh([_.-]|$)/i],
    },
    {
      key: "col",
      label: "col",
      tokens: ["col"],
      patterns: [/(^|[_-])col([_.-]|$)/i],
    },
    {
      key: "normal",
      label: "normal",
      tokens: ["normal"],
      patterns: [/(^|[_-])normal([_.-]|$)/i],
    },
    {
      key: "smile",
      label: "smile",
      tokens: ["smile"],
      patterns: [/(^|[_-])smile([_.-]|$)/i],
    },
    {
      key: "background",
      label: "background",
      tokens: ["background"],
      patterns: [/(^|[_-])background([_.-]|$)/i],
    },
    {
      key: "wink",
      label: "wink",
      tokens: ["wink"],
      patterns: [/(^|[_-])wink([_.-]|$)/i],
    },
    {
      key: "cool",
      label: "cool",
      tokens: ["cool"],
      patterns: [/(^|[_-])cool([_.-]|$)/i],
    },
    {
      key: "sulk",
      label: "sulk",
      tokens: ["sulk"],
      patterns: [/(^|[_-])sulk([_.-]|$)/i],
    },
    {
      key: "doubt",
      label: "doubt",
      tokens: ["doubt"],
      patterns: [/(^|[_-])doubt([_.-]|$)/i],
    },
    {
      key: "painful",
      label: "painful",
      tokens: ["painful"],
      patterns: [/(^|[_-])painful([_.-]|$)/i],
    },
    {
      key: "angry",
      label: "angry",
      tokens: ["angry"],
      patterns: [/(^|[_-])angry([_.-]|$)/i],
    },
    {
      key: "happy",
      label: "happy",
      tokens: ["happy"],
      patterns: [/(^|[_-])happy([_.-]|$)/i],
    },
    {
      key: "noget",
      label: "noget",
      tokens: ["noget"],
      patterns: [/(^|[_-])noget([_.-]|$)/i],
    },
    {
      key: "sad",
      label: "sad",
      tokens: ["sad"],
      patterns: [/(^|[_-])sad([_.-]|$)/i],
    },
    {
      key: "shout",
      label: "shout",
      tokens: ["shout"],
      patterns: [/(^|[_-])shout([_.-]|$)/i],
    },
    {
      key: "sleepy",
      label: "sleepy",
      tokens: ["sleepy"],
      patterns: [/(^|[_-])sleepy([_.-]|$)/i],
    },
    {
      key: "surprise",
      label: "surprise",
      tokens: ["surprise"],
      patterns: [/(^|[_-])surprise([_.-]|$)/i],
    },
    {
      key: "bitter",
      label: "bitter",
      tokens: ["bitter"],
      patterns: [/(^|[_-])bitter([_.-]|$)/i],
    },
    {
      key: "num",
      label: "num",
      tokens: ["num"],
      patterns: [/(^|[_-])num([_.-]|$)/i],
    },
    {
      key: "panic",
      label: "panic",
      tokens: ["panic"],
      patterns: [/(^|[_-])panic([_.-]|$)/i],
    },
    {
      key: "serious",
      label: "serious",
      tokens: ["serious"],
      patterns: [/(^|[_-])serious([_.-]|$)/i],
    },
    {
      key: "square",
      label: "square",
      tokens: ["square"],
      patterns: [/(^|[_-])square([_.-]|$)/i],
    },
    {
      key: "controlmap",
      label: "controlmap",
      tokens: ["controlmap"],
      patterns: [/(^|[_-])controlmap([_.-]|$)/i],
    },
    {
      key: "motivation",
      label: "motivation",
      tokens: ["motivation"],
      patterns: [/(^|[_-])motivation([_.-]|$)/i],
    },
    {
      key: "logo",
      label: "logo",
      tokens: ["logo"],
      patterns: [/(^|[_-])logo([_.-]|$)/i],
    },
    {
      key: "sell",
      label: "sell",
      tokens: ["sell"],
      patterns: [/(^|[_-])sell([_.-]|$)/i],
    },
    {
      key: "standard",
      label: "standard",
      tokens: ["standard"],
      patterns: [/(^|[_-])standard([_.-]|$)/i],
    },
    {
      key: "chara",
      label: "chara",
      tokens: ["chara"],
      patterns: [/(^|[_-])chara([_.-]|$)/i],
    },
    {
      key: "stand",
      label: "stand",
      tokens: ["stand"],
      patterns: [/(^|[_-])stand([_.-]|$)/i],
    },
    {
      key: "shock",
      label: "shock",
      tokens: ["shock"],
      patterns: [/(^|[_-])shock([_.-]|$)/i],
    },
    {
      key: "view",
      label: "view",
      tokens: ["view"],
      patterns: [/(^|[_-])view([_.-]|$)/i],
    },
    {
      key: "download",
      label: "download",
      tokens: ["download"],
      patterns: [/(^|[_-])download([_.-]|$)/i],
    },
    {
      key: "surprised",
      label: "surprised",
      tokens: ["surprised"],
      patterns: [/(^|[_-])surprised([_.-]|$)/i],
    },
    {
      key: "top",
      label: "top",
      tokens: ["top"],
      patterns: [/(^|[_-])top([_.-]|$)/i],
    },
    {
      key: "trouble",
      label: "trouble",
      tokens: ["trouble"],
      patterns: [/(^|[_-])trouble([_.-]|$)/i],
    },
  ],
};

window.FilterUtils = {
  _loadPromise: null,

  tokenizeLabel(label) {
    const tokens = new Set();
    const parts = label
      .toLowerCase()
      .split(/[^a-z0-9]+/g)
      .filter(Boolean);
    parts.forEach((part) => {
      tokens.add(part);
      const trimmed = part.replace(/\d+$/, "");
      if (trimmed && trimmed !== part) {
        tokens.add(trimmed);
      }
    });
    return Array.from(tokens);
  },

  matchLabel(label, filter, tokens) {
    if (!filter) {
      return true;
    }
    let matched = false;
    if (filter.patterns && filter.patterns.length > 0) {
      matched = filter.patterns.some((pattern) => pattern.test(label));
    }
    if (!matched && filter.tokens && filter.tokens.length > 0) {
      const haystack = tokens || FilterUtils.tokenizeLabel(label);
      matched = filter.tokens.some((token) => haystack.includes(token));
    }
    if (!filter.patterns && !filter.tokens) {
      return true;
    }
    return matched;
  },

  loadConfig() {
    if (this._loadPromise) {
      return this._loadPromise;
    }
    this._loadPromise = fetch("/api/filters")
      .then((res) => (res.ok ? res.json() : null))
      .then((data) => {
        if (data) {
          FilterUtils.mergeConfig(data);
        }
      })
      .catch(() => {});
    return this._loadPromise;
  },

  mergeConfig(data) {
    if (!data || !window.FilterConfig) {
      return;
    }
    if (data.media && FilterConfig.media) {
      Object.keys(data.media).forEach((key) => {
        const filter = FilterConfig.media.find((item) => item.key === key);
        if (!filter) {
          return;
        }
        const tokens = Array.isArray(data.media[key]) ? data.media[key] : [];
        filter.tokens = mergeTokens(filter.tokens, tokens);
      });
    }
    if (Array.isArray(data.characters) && FilterConfig.characters) {
      const existing = new Map();
      FilterConfig.characters.forEach((item) => {
        existing.set(item.key, item);
      });
      data.characters.forEach((token) => {
        const key = String(token || "").toLowerCase();
        if (!key) {
          return;
        }
        const entry = existing.get(key);
        if (entry) {
          entry.tokens = mergeTokens(entry.tokens, [key]);
          return;
        }
        const label = String(token);
        const newFilter = {
          key,
          label,
          tokens: [key],
        };
        FilterConfig.characters.push(newFilter);
        existing.set(key, newFilter);
      });
    }
  },
};

function mergeTokens(base, extra) {
  const set = new Set(Array.isArray(base) ? base : []);
  (Array.isArray(extra) ? extra : []).forEach((token) => {
    if (!token) {
      return;
    }
    set.add(String(token).toLowerCase());
  });
  return Array.from(set);
}

File: webui/filters.go
package webui

import (
	"net/http"
	"path/filepath"
	"sort"
	"strings"
	"time"
	"unicode"

	"vertesan/hailstorm/manifest"
)

type AutoFilters struct {
	Media      map[string][]string `json:"media"`
	Characters []string            `json:"characters"`
}

type tokenStat struct {
	total  int
	byKind map[string]int
}

func (s *Server) handleFilters(w http.ResponseWriter, r *http.Request) {
	_ = s.catalog.Reload()
	entries, modTime, _ := s.catalog.Stats()
	cfg := s.autoFilters(entries, modTime)
	writeJSON(w, cfg)
}

func (s *Server) autoFilters(entries []manifest.Entry, modTime time.Time) AutoFilters {
	s.filtersMu.Lock()
	defer s.filtersMu.Unlock()

	if s.filtersOK && modTime.Equal(s.filterMod) {
		return s.filters
	}

	cfg := buildAutoFilters(entries)
	s.filters = cfg
	s.filterMod = modTime
	s.filtersOK = true
	return cfg
}

func buildAutoFilters(entries []manifest.Entry) AutoFilters {
	return AutoFilters{
		Media:      buildMediaTokens(entries),
		Characters: buildCharacterTokens(entries),
	}
}

func buildMediaTokens(entries []manifest.Entry) map[string][]string {
	stats := map[string]*tokenStat{}
	for _, entry := range entries {
		label := assetDisplayName(entry)
		kinds := seedMediaKinds(label, entry.StrTypeCrc)
		if len(kinds) == 0 {
			continue
		}
		tokens := uniqueTokens(tokenizeLabel(label))
		for _, token := range tokens {
			if !isMediaTokenCandidate(token) {
				continue
			}
			stat := stats[token]
			if stat == nil {
				stat = &tokenStat{byKind: map[string]int{}}
				stats[token] = stat
			}
			stat.total++
			for _, kind := range kinds {
				stat.byKind[kind]++
			}
		}
	}

	minCount := mediaTokenThreshold(len(entries))
	result := map[string][]string{}
	for _, kind := range mediaKinds() {
		candidates := []tokenCount{}
		for token, stat := range stats {
			if stat.total < minCount {
				continue
			}
			kindCount := stat.byKind[kind]
			if kindCount < minCount {
				continue
			}
			ratio := float64(kindCount) / float64(stat.total)
			if ratio < 0.68 {
				continue
			}
			candidates = append(candidates, tokenCount{token: token, count: kindCount})
		}
		sortTokens(candidates)
		if len(candidates) > 80 {
			candidates = candidates[:80]
		}
		for _, item := range candidates {
			result[kind] = append(result[kind], item.token)
		}
	}
	return result
}

func buildCharacterTokens(entries []manifest.Entry) []string {
	counts := map[string]int{}
	for _, entry := range entries {
		label := assetDisplayName(entry)
		tokens := tokenizeLabel(label)
		for i := 0; i+1 < len(tokens); i++ {
			if !characterPrefixes[strings.ToLower(tokens[i])] {
				continue
			}
			token := cleanCharacterToken(tokens[i+1])
			if token == "" {
				continue
			}
			counts[token]++
		}
	}

	minCount := characterTokenThreshold(len(entries))
	candidates := []tokenCount{}
	for token, count := range counts {
		if count < minCount {
			continue
		}
		candidates = append(candidates, tokenCount{token: token, count: count})
	}
	sortTokens(candidates)
	if len(candidates) > 60 {
		candidates = candidates[:60]
	}

	out := make([]string, 0, len(candidates))
	for _, item := range candidates {
		out = append(out, item.token)
	}
	return out
}

func seedMediaKinds(label string, entryType string) []string {
	lower := strings.ToLower(label)
	kinds := map[string]bool{}
	for _, suffix := range motionSuffixes {
		if strings.HasSuffix(lower, suffix) {
			kinds["motion"] = true
		}
	}
	for _, suffix := range modelSuffixes {
		if strings.HasSuffix(lower, suffix) {
			kinds["model"] = true
		}
	}

	ext := strings.TrimPrefix(filepath.Ext(lower), ".")
	if ext != "" {
		if kind, ok := mediaExtToKind[ext]; ok {
			kinds[kind] = true
		}
	} else if entryType != "" {
		if kind, ok := mediaExtToKind[strings.ToLower(entryType)]; ok {
			kinds[kind] = true
		}
	}

	if len(kinds) == 0 {
		return nil
	}
	out := make([]string, 0, len(kinds))
	for kind := range kinds {
		out = append(out, kind)
	}
	sort.Strings(out)
	return out
}

func tokenizeLabel(label string) []string {
	raw := strings.FieldsFunc(strings.ToLower(label), func(r rune) bool {
		return !unicode.IsLetter(r) && !unicode.IsDigit(r)
	})
	tokens := make([]string, 0, len(raw))
	for _, token := range raw {
		if token == "" {
			continue
		}
		tokens = append(tokens, token)
		trimmed := strings.TrimRightFunc(token, unicode.IsDigit)
		if trimmed != token && len(trimmed) >= 2 {
			tokens = append(tokens, trimmed)
		}
	}
	return tokens
}

func uniqueTokens(tokens []string) []string {
	seen := map[string]bool{}
	out := make([]string, 0, len(tokens))
	for _, token := range tokens {
		if token == "" || seen[token] {
			continue
		}
		seen[token] = true
		out = append(out, token)
	}
	return out
}

func isMediaTokenCandidate(token string) bool {
	if len(token) < 2 {
		return false
	}
	if mediaTokenStop[token] {
		return false
	}
	return hasLetter(token)
}

func cleanCharacterToken(token string) string {
	token = strings.ToLower(strings.TrimSpace(token))
	if token == "" {
		return ""
	}
	clean := strings.TrimRightFunc(token, unicode.IsDigit)
	if clean == "" {
		return ""
	}
	if !isCharacterToken(clean) {
		return ""
	}
	return clean
}

func isCharacterToken(token string) bool {
	if len(token) < 3 || len(token) > 16 {
		return false
	}
	if characterStop[token] {
		return false
	}
	if hasDigit(token) {
		return false
	}
	return hasLetter(token)
}

func hasLetter(token string) bool {
	for _, r := range token {
		if unicode.IsLetter(r) {
			return true
		}
	}
	return false
}

func hasDigit(token string) bool {
	for _, r := range token {
		if unicode.IsDigit(r) {
			return true
		}
	}
	return false
}

type tokenCount struct {
	token string
	count int
}

func sortTokens(tokens []tokenCount) {
	sort.Slice(tokens, func(i, j int) bool {
		if tokens[i].count == tokens[j].count {
			return tokens[i].token < tokens[j].token
		}
		return tokens[i].count > tokens[j].count
	})
}

func mediaTokenThreshold(total int) int {
	switch {
	case total > 50000:
		return 15
	case total > 15000:
		return 10
	case total > 5000:
		return 6
	default:
		return 3
	}
}

func characterTokenThreshold(total int) int {
	switch {
	case total > 20000:
		return 8
	case total > 5000:
		return 5
	default:
		return 3
	}
}

func mediaKinds() []string {
	return []string{"image", "video", "audio", "model", "motion"}
}

var mediaExtToKind = map[string]string{
	"png":  "image",
	"jpg":  "image",
	"jpeg": "image",
	"webp": "image",
	"gif":  "image",
	"bmp":  "image",
	"tga":  "image",
	"dds":  "image",
	"ktx":  "image",
	"ktx2": "image",
	"tif":  "image",
	"tiff": "image",

	"usm":  "video",
	"mp4":  "video",
	"webm": "video",
	"mov":  "video",
	"m4v":  "video",
	"avi":  "video",
	"mkv":  "video",

	"acb":  "audio",
	"awb":  "audio",
	"hca":  "audio",
	"adx":  "audio",
	"wav":  "audio",
	"ogg":  "audio",
	"mp3":  "audio",
	"flac": "audio",
	"aac":  "audio",
	"m4a":  "audio",
	"opus": "audio",
	"wma":  "audio",

	"glb":  "model",
	"gltf": "model",
	"fbx":  "model",
	"obj":  "model",
	"dae":  "model",

	"anim":       "motion",
	"controller": "motion",
	"mot":        "motion",
}

var motionSuffixes = []string{
	".anim.assetbundle",
	".controller.assetbundle",
	".motion.assetbundle",
}

var modelSuffixes = []string{
	".playable.assetbundle",
	".model.assetbundle",
	".mesh.assetbundle",
}

var mediaTokenStop = map[string]bool{
	"asset":       true,
	"assets":      true,
	"assetbundle": true,
	"bundle":      true,
	"data":        true,
	"common":      true,
	"shared":      true,
	"main":        true,
	"pack":        true,
	"res":         true,
	"resource":    true,
	"resources":   true,
	"cache":       true,
	"plain":       true,
}

var characterPrefixes = map[string]bool{
	"chara":     true,
	"character": true,
	"char":      true,
	"chr":       true,
	"card":      true,
	"idol":      true,
	"unit":      true,
	"dress":     true,
	"costume":   true,
	"live":      true,
	"profile":   true,
	"portrait":  true,
	"face":      true,
	"icon":      true,
	"voice":     true,
	"vo":        true,
	"model":     true,
	"motion":    true,
}

var characterStop = map[string]bool{
	"common":        true,
	"default":       true,
	"main":          true,
	"asset":         true,
	"assets":        true,
	"skill":         true,
	"middle":        true,
	"full":          true,
	"half":          true,
	"season":        true,
	"adv":           true,
	"hime":          true,
	"item":          true,
	"get":           true,
	"frame":         true,
	"specialappeal": true,
	"area":          true,
	"laugh":         true,
	"col":           true,
	"normal":        true,
	"smile":         true,
	"background":    true,
	"wink":          true,
	"cool":          true,
	"sulk":          true,
	"doubt":         true,
	"painful":       true,
	"angry":         true,
	"happy":         true,
	"noget":         true,
	"sad":           true,
	"shout":         true,
	"sleepy":        true,
	"surprise":      true,
	"bitter":        true,
	"num":           true,
	"panic":         true,
	"serious":       true,
	"square":        true,
	"controlmap":    true,
	"motivation":    true,
	"logo":          true,
	"sell":          true,
	"standard":      true,
	"chara":         true,
	"stand":         true,
	"shock":         true,
	"view":          true,
	"download":      true,
	"surprised":     true,
	"top":           true,
	"trouble":       true,
	"icon":          true,
	"face":          true,
	"card":          true,
	"voice":         true,
	"vo":            true,
	"bgm":           true,
	"se":            true,
	"music":         true,
	"story":         true,
	"model":         true,
	"motion":        true,
	"event":         true,
	"shop":          true,
	"banner":        true,
	"title":         true,
	"scene":         true,
	"sprite":        true,
	"texture":       true,
	"image":         true,
	"movie":         true,
	"video":         true,
	"sound":         true,
	"stage":         true,
	"effect":        true,
	"effects":       true,
	"master":        true,
	"sample":        true,
	"test":          true,
	"dummy":         true,
	"tutorial":      true,
}
